File: kuber-client-csharp/examples-standalone/ActiveMqRequestResponseTest.cs
/*
 * Copyright © 2025-2030, All Rights Reserved
 * Ashutosh Sinha | Email: ajsinha@gmail.com
 *
 * Kuber ActiveMQ Request/Response Test Client (v1.7.5) - C#/.NET
 * 
 * Tests the Kuber Request/Response messaging feature via Apache ActiveMQ.
 * Publishes cache operation requests and subscribes to responses.
 *
 * Prerequisites:
 * - ActiveMQ running on localhost:61616
 * - Kuber server running with messaging enabled
 * - Install NuGet packages:
 *     dotnet add package Apache.NMS
 *     dotnet add package Apache.NMS.ActiveMQ
 */

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Threading;
using System.Threading.Tasks;
using Apache.NMS;
using Apache.NMS.ActiveMQ;

namespace Kuber.Client.Examples
{
    /// <summary>
    /// Test client for Kuber's ActiveMQ-based Request/Response messaging.
    /// 
    /// Demonstrates how to:
    /// - Publish cache operation requests to ActiveMQ queues
    /// - Subscribe to and receive responses
    /// - Correlate requests with responses using message IDs
    /// </summary>
    public class ActiveMqRequestResponseTest
    {
        // Configuration
        private const string BrokerUri = "activemq:tcp://localhost:61616";
        private const string Username = "admin";
        private const string Password = "admin";
        private const string RequestQueue = "ccs_cache_request";
        private const string ResponseQueue = "ccs_cache_response";
        
        // Your Kuber API key (replace with actual key)
        private const string ApiKey = "kub_your_api_key_here";
        
        // Test region
        private const string TestRegion = "default";
        
        // NMS objects
        private IConnectionFactory? _factory;
        private IConnection? _connection;
        private ISession? _session;
        private IMessageProducer? _producer;
        private IMessageConsumer? _consumer;
        
        // State
        private readonly CancellationTokenSource _cts = new();
        private readonly ConcurrentDictionary<string, TaskCompletionSource<JsonNode>> _pendingRequests = new();
        private readonly List<JsonNode> _receivedResponses = new();
        
        /// <summary>
        /// Main entry point.
        /// </summary>
        public static async Task Main(string[] args)
        {
            Console.WriteLine("╔════════════════════════════════════════════════════════════════════╗");
            Console.WriteLine("║  Kuber ActiveMQ Request/Response Test Client (C#/.NET)             ║");
            Console.WriteLine("║  Tests cache operations via ActiveMQ messaging                     ║");
            Console.WriteLine("╚════════════════════════════════════════════════════════════════════╝");
            Console.WriteLine();
            
            var test = new ActiveMqRequestResponseTest();
            try
            {
                await test.RunAsync();
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Error: {ex.Message}");
                Console.Error.WriteLine(ex.StackTrace);
            }
        }
        
        /// <summary>
        /// Run the test suite.
        /// </summary>
        public async Task RunAsync()
        {
            try
            {
                // Initialize NMS connection
                InitConnection();
                
                // Start response listener
                StartResponseListener();
                
                // Give consumer time to connect
                await Task.Delay(2000);
                
                // Run tests
                await RunTestSuiteAsync();
                
                // Wait for responses
                Console.WriteLine("\nWaiting for responses (10 seconds)...");
                await Task.Delay(10000);
            }
            finally
            {
                Shutdown();
                PrintSummary();
            }
        }
        
        /// <summary>
        /// Initialize NMS connection to ActiveMQ.
        /// </summary>
        private void InitConnection()
        {
            Console.WriteLine($"Connecting to ActiveMQ at {BrokerUri}...");
            
            _factory = new ConnectionFactory(BrokerUri);
            _connection = _factory.CreateConnection(Username, Password);
            _connection.Start();
            
            _session = _connection.CreateSession(AcknowledgementMode.AutoAcknowledge);
            
            // Create producer for request queue
            var requestDest = _session.GetQueue(RequestQueue);
            _producer = _session.CreateProducer(requestDest);
            _producer.DeliveryMode = MsgDeliveryMode.Persistent;
            
            // Create consumer for response queue
            var responseDest = _session.GetQueue(ResponseQueue);
            _consumer = _session.CreateConsumer(responseDest);
            
            Console.WriteLine("✓ Connected to ActiveMQ");
            Console.WriteLine($"  Request queue: {RequestQueue}");
            Console.WriteLine($"  Response queue: {ResponseQueue}");
        }
        
        /// <summary>
        /// Start listening for responses.
        /// </summary>
        private void StartResponseListener()
        {
            _consumer!.Listener += message =>
            {
                try
                {
                    if (message is ITextMessage textMessage)
                    {
                        var responseText = textMessage.Text;
                        var response = JsonNode.Parse(responseText);
                        if (response == null) return;
                        
                        // Kuber response structure is nested:
                        // {
                        //   "request_receive_timestamp": "...",
                        //   "response_time": "...",
                        //   "processing_time_ms": 123,
                        //   "request": { original request with message_id, operation, etc. },
                        //   "response": { "success": true/false, "result": ..., "error": "..." }
                        // }
                        var requestNode = response["request"];
                        var responseNode = response["response"];
                        
                        var messageId = requestNode?["message_id"]?.GetValue<string>() ?? "unknown";
                        var success = responseNode?["success"]?.GetValue<bool>() ?? false;
                        var operation = requestNode?["operation"]?.GetValue<string>() ?? "unknown";
                        var processingTime = response["processing_time_ms"]?.GetValue<long>() ?? 0;
                        
                        Console.WriteLine();
                        Console.WriteLine("════════════════════════════════════════════════════════════");
                        Console.WriteLine("📥 RESPONSE RECEIVED");
                        Console.WriteLine($"   Message ID: {messageId}");
                        Console.WriteLine($"   Success: {success}");
                        Console.WriteLine($"   Operation: {operation}");
                        Console.WriteLine($"   Processing Time: {processingTime}ms");
                        
                        if (success)
                        {
                            var resultValue = responseNode?["result"]?.ToJsonString();
                            if (!string.IsNullOrEmpty(resultValue))
                            {
                                if (resultValue.Length > 100)
                                {
                                    Console.WriteLine($"   Result: {resultValue[..100]}...");
                                }
                                else
                                {
                                    Console.WriteLine($"   Result: {resultValue}");
                                }
                            }
                        }
                        else
                        {
                            var errorCode = responseNode?["error_code"]?.GetValue<string>() ?? "UNKNOWN";
                            var error = responseNode?["error"]?.GetValue<string>() ?? "Unknown error";
                            Console.WriteLine($"   Error Code: {errorCode}");
                            Console.WriteLine($"   Error: {error}");
                        }
                        Console.WriteLine("════════════════════════════════════════════════════════════");
                        
                        lock (_receivedResponses)
                        {
                            _receivedResponses.Add(response);
                        }
                        
                        // Complete pending task if exists
                        if (_pendingRequests.TryRemove(messageId, out var tcs))
                        {
                            tcs.SetResult(response);
                        }
                    }
                }
                catch (Exception ex)
                {
                    Console.Error.WriteLine($"Error processing response: {ex.Message}");
                }
            };
            
            Console.WriteLine("✓ Response listener started");
        }
        
        /// <summary>
        /// Create a request message.
        /// </summary>
        private JsonObject CreateRequest(string operation, string region, string key,
            string? value = null, int? ttl = null)
        {
            var messageId = $"req-{Guid.NewGuid().ToString()[..12]}";
            
            var request = new JsonObject
            {
                ["api_key"] = ApiKey,
                ["message_id"] = messageId,
                ["operation"] = operation.ToUpperInvariant(),
                ["region"] = region,
                ["key"] = key
            };
            
            if (!string.IsNullOrEmpty(value))
            {
                request["value"] = value;
            }
            
            if (ttl.HasValue)
            {
                request["ttl"] = ttl.Value;
            }
            
            return request;
        }
        
        /// <summary>
        /// Publish a request to ActiveMQ.
        /// </summary>
        private TaskCompletionSource<JsonNode> PublishRequest(JsonObject request)
        {
            var messageId = request["message_id"]!.GetValue<string>();
            var requestJson = request.ToJsonString();
            
            var tcs = new TaskCompletionSource<JsonNode>();
            _pendingRequests[messageId] = tcs;
            
            try
            {
                var message = _session!.CreateTextMessage(requestJson);
                message.NMSCorrelationID = messageId;
                _producer!.Send(message);
                
                Console.WriteLine();
                Console.WriteLine("📤 REQUEST SENT");
                Console.WriteLine($"   Message ID: {messageId}");
                Console.WriteLine($"   Operation: {request["operation"]}");
                Console.WriteLine($"   Region: {request["region"]}");
                Console.WriteLine($"   Key: {request["key"]}");
                Console.WriteLine($"   Queue: {RequestQueue}");
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"✗ Failed to send: {ex.Message}");
                tcs.SetException(ex);
            }
            
            return tcs;
        }
        
        /// <summary>
        /// Run the test suite.
        /// </summary>
        private async Task RunTestSuiteAsync()
        {
            Console.WriteLine("\n--- Test 1: SET Operation ---");
            var setRequest = CreateRequest("SET", TestRegion,
                "activemq-test:user:1001",
                JsonSerializer.Serialize(new { name = "John Doe", email = "john@example.com", timestamp = DateTime.UtcNow }),
                3600);
            PublishRequest(setRequest);
            await Task.Delay(1000);
            
            Console.WriteLine("\n--- Test 2: GET Operation ---");
            var getRequest = CreateRequest("GET", TestRegion, "activemq-test:user:1001");
            PublishRequest(getRequest);
            await Task.Delay(1000);
            
            Console.WriteLine("\n--- Test 3: EXISTS Operation ---");
            var existsRequest = CreateRequest("EXISTS", TestRegion, "activemq-test:user:1001");
            PublishRequest(existsRequest);
            await Task.Delay(1000);
            
            Console.WriteLine("\n--- Test 4: SET Another Key ---");
            var setRequest2 = CreateRequest("SET", TestRegion,
                "activemq-test:user:1002",
                JsonSerializer.Serialize(new { name = "Jane Smith", role = "admin" }));
            PublishRequest(setRequest2);
            await Task.Delay(1000);
            
            Console.WriteLine("\n--- Test 5: KEYS Operation ---");
            var keysRequest = CreateRequest("KEYS", TestRegion, "activemq-test:*");
            PublishRequest(keysRequest);
            await Task.Delay(1000);
            
            Console.WriteLine("\n--- Test 6: DELETE Operation ---");
            var deleteRequest = CreateRequest("DELETE", TestRegion, "activemq-test:user:1002");
            PublishRequest(deleteRequest);
            await Task.Delay(1000);
            
            Console.WriteLine("\n--- Test 7: GET Non-existent Key ---");
            var getMissRequest = CreateRequest("GET", TestRegion, "activemq-test:nonexistent:key");
            PublishRequest(getMissRequest);
            await Task.Delay(1000);
            
            Console.WriteLine("\n✓ All requests sent");
        }
        
        /// <summary>
        /// Shutdown connections.
        /// </summary>
        private void Shutdown()
        {
            Console.WriteLine("\nShutting down...");
            _cts.Cancel();
            
            try
            {
                _consumer?.Close();
                _producer?.Close();
                _session?.Close();
                _connection?.Close();
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Error closing connections: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Print test summary.
        /// </summary>
        private void PrintSummary()
        {
            Console.WriteLine();
            Console.WriteLine("════════════════════════════════════════════════════════════");
            Console.WriteLine("TEST SUMMARY");
            Console.WriteLine("════════════════════════════════════════════════════════════");
            
            lock (_receivedResponses)
            {
                Console.WriteLine($"Total responses received: {_receivedResponses.Count}");
            }
            Console.WriteLine($"Pending requests (no response): {_pendingRequests.Count}");
            
            if (!_pendingRequests.IsEmpty)
            {
                Console.WriteLine("\nRequests without responses:");
                foreach (var messageId in _pendingRequests.Keys)
                {
                    Console.WriteLine($"  - {messageId}");
                }
            }
            
            int successCount = 0;
            lock (_receivedResponses)
            {
                foreach (var r in _receivedResponses)
                {
                    var responseNode = r["response"];
                    if (responseNode?["success"]?.GetValue<bool>() == true)
                    {
                        successCount++;
                    }
                }
            }
            
            var errorCount = _receivedResponses.Count - successCount;
            
            Console.WriteLine($"\nSuccessful operations: {successCount}");
            Console.WriteLine($"Failed operations: {errorCount}");
            Console.WriteLine("════════════════════════════════════════════════════════════");
        }
    }
}

File: kuber-client-java/examples/ActiveMqRequestResponseTest.java
/*
 * Copyright © 2025-2030, All Rights Reserved
 * Ashutosh Sinha | Email: ajsinha@gmail.com
 *
 * Kuber ActiveMQ Request/Response Test Client (v1.7.9)
 * 
 * Tests the Kuber Request/Response messaging feature via Apache ActiveMQ.
 * Publishes cache operation requests and subscribes to responses.
 *
 * Prerequisites:
 * - ActiveMQ running on localhost:61616
 * - Kuber server running with messaging enabled
 * - activemq-client dependency in pom.xml:
 *     <dependency>
 *         <groupId>org.apache.activemq</groupId>
 *         <artifactId>activemq-client</artifactId>
 *         <version>5.18.3</version>
 *     </dependency>
 */
package com.kuber.client.examples;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.apache.activemq.ActiveMQConnectionFactory;

import javax.jms.*;
import java.time.Instant;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Test client for Kuber's ActiveMQ-based Request/Response messaging.
 * 
 * <p>Demonstrates how to:
 * <ul>
 *   <li>Publish cache operation requests to ActiveMQ queues</li>
 *   <li>Subscribe to and receive responses</li>
 *   <li>Correlate requests with responses using message IDs</li>
 * </ul>
 * 
 * @version 1.7.9
 */
public class ActiveMqRequestResponseTest {
    
    // Configuration
    private static final String BROKER_URL = "tcp://localhost:61616";
    private static final String USERNAME = "admin";
    private static final String PASSWORD = "admin";
    private static final String REQUEST_QUEUE = "ccs_cache_request";
    private static final String RESPONSE_QUEUE = "ccs_cache_response";
    
    // Your Kuber API key (replace with actual key)
    private static final String API_KEY = "kub_your_api_key_here";
    
    // Test region
    private static final String TEST_REGION = "default";
    
    // JSON mapper
    private final ObjectMapper objectMapper = new ObjectMapper();
    
    // JMS objects
    private Connection connection;
    private Session session;
    private MessageProducer producer;
    private MessageConsumer consumer;
    
    // State
    private final AtomicBoolean running = new AtomicBoolean(true);
    private final Map<String, CompletableFuture<JsonNode>> pendingRequests = new ConcurrentHashMap<>();
    private final List<JsonNode> receivedResponses = new CopyOnWriteArrayList<>();
    
    /**
     * Main entry point.
     */
    public static void main(String[] args) {
        System.out.println("╔════════════════════════════════════════════════════════════════════╗");
        System.out.println("║  Kuber ActiveMQ Request/Response Test Client (Java)                ║");
        System.out.println("║  Tests cache operations via ActiveMQ messaging                     ║");
        System.out.println("╚════════════════════════════════════════════════════════════════════╝");
        System.out.println();
        
        ActiveMqRequestResponseTest test = new ActiveMqRequestResponseTest();
        try {
            test.run();
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }
    
    /**
     * Run the test suite.
     */
    public void run() throws Exception {
        try {
            // Initialize JMS connection
            initConnection();
            
            // Start response listener
            startResponseListener();
            
            // Give consumer time to connect
            Thread.sleep(2000);
            
            // Run tests
            runTestSuite();
            
            // Wait for responses
            System.out.println("\nWaiting for responses (10 seconds)...");
            Thread.sleep(10000);
            
        } finally {
            shutdown();
            printSummary();
        }
    }
    
    /**
     * Initialize JMS connection to ActiveMQ.
     */
    private void initConnection() throws JMSException {
        System.out.println("Connecting to ActiveMQ at " + BROKER_URL + "...");
        
        ActiveMQConnectionFactory factory = new ActiveMQConnectionFactory(BROKER_URL);
        factory.setUserName(USERNAME);
        factory.setPassword(PASSWORD);
        
        connection = factory.createConnection();
        connection.start();
        
        session = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
        
        // Create producer for request queue
        javax.jms.Queue requestQueue = session.createQueue(REQUEST_QUEUE);
        producer = session.createProducer(requestQueue);
        producer.setDeliveryMode(DeliveryMode.PERSISTENT);
        
        // Create consumer for response queue
        javax.jms.Queue responseQueue = session.createQueue(RESPONSE_QUEUE);
        consumer = session.createConsumer(responseQueue);
        
        System.out.println("✓ Connected to ActiveMQ");
        System.out.println("  Request queue: " + REQUEST_QUEUE);
        System.out.println("  Response queue: " + RESPONSE_QUEUE);
    }
    
    /**
     * Start listening for responses.
     */
    private void startResponseListener() throws JMSException {
        consumer.setMessageListener(message -> {
            try {
                if (message instanceof TextMessage) {
                    String responseText = ((TextMessage) message).getText();
                    JsonNode response = objectMapper.readTree(responseText);
                    
                    // Kuber response structure is nested:
                    // {
                    //   "request_receive_timestamp": "...",
                    //   "response_time": "...",
                    //   "processing_time_ms": 123,
                    //   "request": { original request with message_id, operation, etc. },
                    //   "response": { "success": true/false, "result": ..., "error": "..." }
                    // }
                    JsonNode requestNode = response.get("request");
                    JsonNode responseNode = response.get("response");
                    
                    String messageId = requestNode != null && requestNode.has("message_id") ? 
                        requestNode.get("message_id").asText() : "unknown";
                    boolean success = responseNode != null && responseNode.has("success") && 
                        responseNode.get("success").asBoolean();
                    String operation = requestNode != null && requestNode.has("operation") ?
                        requestNode.get("operation").asText() : "unknown";
                    long processingTime = response.has("processing_time_ms") ?
                        response.get("processing_time_ms").asLong() : 0;
                    
                    System.out.println("\n════════════════════════════════════════════════════════════");
                    System.out.println("📥 RESPONSE RECEIVED");
                    System.out.println("   Message ID: " + messageId);
                    System.out.println("   Success: " + success);
                    System.out.println("   Operation: " + operation);
                    System.out.println("   Processing Time: " + processingTime + "ms");
                    
                    if (success) {
                        if (responseNode != null && responseNode.has("result")) {
                            String result = responseNode.get("result").toString();
                            if (result.length() > 100) {
                                System.out.println("   Result: " + result.substring(0, 100) + "...");
                            } else {
                                System.out.println("   Result: " + result);
                            }
                        }
                    } else {
                        if (responseNode != null) {
                            String errorCode = responseNode.has("error_code") ?
                                responseNode.get("error_code").asText() : "UNKNOWN";
                            String error = responseNode.has("error") ?
                                responseNode.get("error").asText() : "Unknown error";
                            System.out.println("   Error Code: " + errorCode);
                            System.out.println("   Error: " + error);
                        }
                    }
                    System.out.println("════════════════════════════════════════════════════════════");
                    
                    receivedResponses.add(response);
                    
                    // Complete pending future if exists
                    CompletableFuture<JsonNode> future = pendingRequests.remove(messageId);
                    if (future != null) {
                        future.complete(response);
                    }
                }
            } catch (Exception e) {
                System.err.println("Error processing response: " + e.getMessage());
            }
        });
        
        System.out.println("✓ Response listener started");
    }
    
    /**
     * Create a request message.
     */
    private ObjectNode createRequest(String operation, String region, String key, 
                                      String value, Integer ttl) {
        ObjectNode request = objectMapper.createObjectNode();
        
        String messageId = "req-" + UUID.randomUUID().toString().substring(0, 12);
        
        request.put("api_key", API_KEY);
        request.put("message_id", messageId);
        request.put("operation", operation.toUpperCase());
        request.put("region", region);
        request.put("key", key);
        
        if (value != null) {
            request.put("value", value);
        }
        
        if (ttl != null) {
            request.put("ttl", ttl);
        }
        
        return request;
    }
    
    /**
     * Publish a request to ActiveMQ.
     */
    private CompletableFuture<JsonNode> publishRequest(ObjectNode request) throws Exception {
        String messageId = request.get("message_id").asText();
        String requestJson = objectMapper.writeValueAsString(request);
        
        CompletableFuture<JsonNode> responseFuture = new CompletableFuture<>();
        pendingRequests.put(messageId, responseFuture);
        
        TextMessage message = session.createTextMessage(requestJson);
        message.setJMSCorrelationID(messageId);
        producer.send(message);
        
        System.out.println("\n📤 REQUEST SENT");
        System.out.println("   Message ID: " + messageId);
        System.out.println("   Operation: " + request.get("operation").asText());
        System.out.println("   Region: " + request.get("region").asText());
        System.out.println("   Key: " + request.get("key").asText());
        System.out.println("   Queue: " + REQUEST_QUEUE);
        
        return responseFuture;
    }
    
    /**
     * Run the test suite.
     */
    private void runTestSuite() throws Exception {
        System.out.println("\n--- Test 1: SET Operation ---");
        ObjectNode setRequest = createRequest("SET", TEST_REGION, 
            "activemq-test:user:1001",
            "{\"name\":\"John Doe\",\"email\":\"john@example.com\",\"timestamp\":\"" + 
                Instant.now() + "\"}",
            3600);
        publishRequest(setRequest);
        Thread.sleep(1000);
        
        System.out.println("\n--- Test 2: GET Operation ---");
        ObjectNode getRequest = createRequest("GET", TEST_REGION, 
            "activemq-test:user:1001", null, null);
        publishRequest(getRequest);
        Thread.sleep(1000);
        
        System.out.println("\n--- Test 3: EXISTS Operation ---");
        ObjectNode existsRequest = createRequest("EXISTS", TEST_REGION, 
            "activemq-test:user:1001", null, null);
        publishRequest(existsRequest);
        Thread.sleep(1000);
        
        System.out.println("\n--- Test 4: SET Another Key ---");
        ObjectNode setRequest2 = createRequest("SET", TEST_REGION, 
            "activemq-test:user:1002",
            "{\"name\":\"Jane Smith\",\"role\":\"admin\"}", null);
        publishRequest(setRequest2);
        Thread.sleep(1000);
        
        System.out.println("\n--- Test 5: KEYS Operation ---");
        ObjectNode keysRequest = createRequest("KEYS", TEST_REGION, 
            "activemq-test:*", null, null);
        publishRequest(keysRequest);
        Thread.sleep(1000);
        
        System.out.println("\n--- Test 6: DELETE Operation ---");
        ObjectNode deleteRequest = createRequest("DELETE", TEST_REGION, 
            "activemq-test:user:1002", null, null);
        publishRequest(deleteRequest);
        Thread.sleep(1000);
        
        System.out.println("\n--- Test 7: GET Non-existent Key ---");
        ObjectNode getMissRequest = createRequest("GET", TEST_REGION, 
            "activemq-test:nonexistent:key", null, null);
        publishRequest(getMissRequest);
        Thread.sleep(1000);
        
        System.out.println("\n✓ All requests sent");
    }
    
    /**
     * Shutdown connections.
     */
    private void shutdown() {
        System.out.println("\nShutting down...");
        running.set(false);
        
        try {
            if (consumer != null) consumer.close();
            if (producer != null) producer.close();
            if (session != null) session.close();
            if (connection != null) connection.close();
        } catch (JMSException e) {
            System.err.println("Error closing connections: " + e.getMessage());
        }
    }
    
    /**
     * Print test summary.
     */
    private void printSummary() {
        System.out.println("\n════════════════════════════════════════════════════════════");
        System.out.println("TEST SUMMARY");
        System.out.println("════════════════════════════════════════════════════════════");
        System.out.println("Total responses received: " + receivedResponses.size());
        System.out.println("Pending requests (no response): " + pendingRequests.size());
        
        if (!pendingRequests.isEmpty()) {
            System.out.println("\nRequests without responses:");
            for (String messageId : pendingRequests.keySet()) {
                System.out.println("  - " + messageId);
            }
        }
        
        long successCount = receivedResponses.stream()
            .filter(r -> {
                JsonNode responseNode = r.get("response");
                return responseNode != null && responseNode.has("success") && responseNode.get("success").asBoolean();
            })
            .count();
        long errorCount = receivedResponses.size() - successCount;
        
        System.out.println("\nSuccessful operations: " + successCount);
        System.out.println("Failed operations: " + errorCount);
        System.out.println("════════════════════════════════════════════════════════════");
    }
}
